### ISSUE TITLE
S2RegionTermIndexer fails with gcc undefined sanitizer

### ISSUE THREAD

**Issue by benhannel** [google/s2geometry#85]
`S2RegionTermIndexer.GetIndexTerms` fails with gcc undefined sanitizer on. To reproduce, try to run this code with the gcc undefined sanitizer:

```
  // Create an indexer that uses up to 10 cells to approximate each region.
  S2RegionTermIndexer::Options options;
  options.set_max_cells(10);
  S2RegionTermIndexer indexer(options);

  // For this example, we index a disc-shaped region with a 10km radius.
  S2LatLng center = S2LatLng::FromDegrees(44.1, -56.235);
  S1Angle radius = S2Earth::ToAngle(util::units::Kilometers(10.0));
  S2Cap cap(center.ToPoint(), radius);

  // Add the terms for this disc-shaped region to the index.
  for (const auto& term : indexer.GetIndexTerms(cap, "")) {
    std::cout << "Index term " << term << std::endl;
  }
```

Undefined sanitizer output:
```
/usr/local/include/s2/s2cell.h:72:32: runtime error: member access within address 0x55f6f2882230 which does not point to an object of type 'S2Cell'
0x55f6f2882230: note: object has invalid vptr
 00 00 00 00  a0 7c 7a 26 3c 7f 00 00  02 09 03 00 00 00 00 00  64 3e 4b 00 00 00 00 00  55 55 55 55
              ^~~~~~~~~~~~~~~~~~~~~~~
              invalid vptr
*** Aborted at 1568742703 (Unix time, try 'date -d @1568742703') ***
*** Signal 11 (SIGSEGV) (0x71) received by PID 40 (pthread TID 0x7f3c32a8e000) (linux TID 40) (code: address not mapped to object), stack trace: ***
    @ 00007f3c2ac602a3 folly::symbolizer::(anonymous namespace)::signalHandler(int, siginfo_t*, void*)
                       /tmp/folly/folly/experimental/symbolizer/SignalHandler.cpp:431
    @ 00007f3c263faf1f (unknown)
    @ 00007f3c277773fd __dynamic_cast
    @ 00007f3c269d1196 (unknown)
    @ 00007f3c269cfdf5 (unknown)
    @ 00007f3c269d05a0 __ubsan_handle_dynamic_type_cache_miss
    @ 00007f3c3003425a S2Cell::id() const
                       /usr/local/include/s2/s2cell.h:72
                       -> /opt/rockset/rs/src/cpp/compiler/v2/rbo.cc
    @ 00007f3c280ffe92 S2RegionCoverer::AddCandidate(S2RegionCoverer::Candidate*)
    @ 00007f3c28100b95 S2RegionCoverer::GetCoveringInternal(S2Region const&)
    @ 00007f3c28101018 S2RegionCoverer::GetCovering(S2Region const&)
    @ 00007f3c280fd0e0 S2RegionTermIndexer::GetIndexTerms[abi:cxx11](S2Region const&, absl::string_view)
```

This may be caused by flexible length array members. [S2RegionCoverer](https://github.com/google/s2geometry/blob/master/src/s2/s2region_coverer.h#L251) uses a flexible array member in the candidate struct, which is [not valid C++](https://en.wikipedia.org/wiki/Flexible_array_member). [GCC supports them](https://www.geeksforgeeks.org/variable-length-arrays-in-c-and-c/) with a custom extension, but they aren't in the standard.

**Comment by jmr** [comment]
I don't think this is because of the zero-sized array. You could try by changing it to size 1.

I wasn't able to reproduce, but could you try with this patch?

```
@@ -96,8 +96,9 @@ S2RegionCoverer::Candidate* S2RegionCove
   if (!is_terminal) {
     children_size = sizeof(Candidate*) << max_children_shift();
   }
-  Candidate* candidate = static_cast<Candidate*>(
-      ::operator new(sizeof(Candidate) + children_size));
+  void* const candidate_storage = ::operator new (
+      sizeof(Candidate) + children_size, std::align_val_t{alignof(Candidate)});                                                                             
+  Candidate* candidate = new (candidate_storage) Candidate;
   candidate->cell = cell;
   candidate->is_terminal = is_terminal;
   candidate->num_children = 0;
@@ -115,7 +116,8 @@ void S2RegionCoverer::DeleteCandidate(Ca
     for (int i = 0; i < candidate->num_children; ++i)
       DeleteCandidate(candidate->children[i], true);
   }
-  ::operator delete(candidate);
+  candidate->~Candidate();
+  ::operator delete (candidate, std::align_val_t{alignof(Candidate)});
```

Line numbers may be slightly different since this is against a newer version, but it should still apply.

**Comment by jmr** [comment]
I merged #124 . Let me know if this helps, but I was never able to reproduce this.

@edzer

**Comment by benhannel** [comment]
The issue no longer reproduces. Thank you so much!
